tmixer/man/smixer.1
.\"
.\" smixer.1 - a simple interface to /dev/mixer
.\"
.\" Copyright (C) 2000 David Johnson
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2, or (at your option)
.\" any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software
.\" Foundation, 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA. */
.TH SMIXER 1 "v1.0.3 - 2003-06-09" "smixer"
.SH NAME
smixer \- a simple interface to /dev/mixer
.SH SYNOPSIS
.nf
smixer [-a] -p           (print settings)
smixer [-a] --set [file] (set settings from file or stdin)
.fi
.SH DESCRIPTION
smixer is a command-line and scriptable program to control and display
the mixer volume levels on a sound card in linux. smixer is ideal
for setting mixer settings on startup or to switch between multiple
configurations quickly.
.SH OPTIONS
smixer may be invoked with the following command-line options:
.TP
.B -a
When printing current settings include all devices.
.TP
.B -p
Print out the current mixer settings, then exit.
.TP
.B --set \fIfile\fP
Process commands from
.I file,
then exit.
.TP
.B --set
Process commands from stdin (interactive or from a pipe), then exit.
.PP

.SH
Command Synopsis
These commands can be issued through an input file or from stdin when smixer is started in the
.B --set
mode

.TP
.B show
Print out the current mixer settings.  This is equivilant to running 
smixer with the 
.B -p
argument.

.TP
.B vol \fIname\fP \fIvalue\fP
Set the volume for a specific input or output device.
.I name
is the name of the device as shown in the
.B show
command.
.I value
is the percent volume to set or ``-'' to decrease 5% or ``+'' to increase
5%

.TP
.B recsrc \fIname\fP
Set the recording source to a specific input device.
.I name
is the name of an input device as shown in the
.B show
command.  Only sources that are capable of input can be set.

.TP
.B help
Print out a short description of the commands

.TP
.B quit
End command processing and exit the program.

.SH
Example Commands

.TP
vol Vol 70
Sets the master volume to 70%
.TP
vol CD 0
Sets the CD input volume to 0%
.TP
vol pcm +
Increases the pcm volume 5%
.TP
recsrc Mic
Sets the recording source to the Mic input device

.SH
Example Output

.PP
 $ smixer -p
 ID: Name       Act  Rec  Channels     Volume
 --  --------   ---- ---- -------- ----------
  0: Vol        Yes  Yes  Stereo   25700 100%
  1: Bass       Yes       Stereo   12850  50%
  2: Trebl      Yes       Stereo   12850  50%
  4: Pcm        Yes  Yes  Stereo   22616  88%
  5: Spkr       Yes       Mono         0   0%
  6: Line       Yes  Yes  Stereo       0   0%
  7: Mic        Yes  Yes* Mono         0   0%
  8: CD         Yes  Yes  Stereo       0   0%
 11: Rec        Yes       Stereo       0   0%
 13: OGain      Yes  Yes  Stereo    1285   5%
 14: Line1      Yes  Yes  Stereo       0   0%
 15: Line2      Yes       Stereo   25700 100%
 16: Line3      Yes       Stereo   25700 100%
 17: Digital1   Yes       Stereo       0   0%
 18: Digital2   Yes       Stereo   25700 100%
 20: PhoneIn    Yes  Yes  Mono         0   0%
 24: Monitor         Yes  Mono         0   0%


.SH
BUGS
.PP
E-mail bug reports to
.BR smixer@centerclick.org .
smixer has been tested on several linux systems and might work
in other OSS supported OS's aswell. Newer versions can be obtained from
.BR http://centerclick.org/programs/smixer/


ChangeLog
2004-11-14  Philip Lowman  <philip@yhbt.com>

	* UPGRADE: Howto from 1.20 to 1.21

2004-11-14  Philip Lowman  <philip@yhbt.com>

	* html/admin/admin_daemon.php: Some issues with newer code where -INT
	isn't good enough and we have to do a -KILL (with tunezd.pl & tunezd.php)

2004-11-14  Philip Lowman  <philip@yhbt.com>

	* INSTALL: Improved wording

2004-11-14  Philip Lowman  <philip@yhbt.com>

	* config.inc.php: Added non-mandatory option which allows the local-php
	mode to choose which random algorithm to use.

2004-11-13  Philip Lowman  <philip@yhbt.com>

	* system/classes/PQueue.class.php: Makes the new random selection SQL work
	in PHP but requires a config.inc.php variable to be set.
	* tunezd.php: Now supports the new PQueue.class.php.  Also fixed a bug
	with ogg vorbis playback not working!

2004-11-13  Adi Linden  <alinden@users.sourceforge.net>
	[PATCH]

	* system/classes/song.class.php: OGG vorbis track numbers & genre IDs are
	supported now (thanks to Adi Linden's patch
	https://sourceforge.net/forum/forum.php?thread_id=1177457&forum_id=6592)
	
2004-11-13  Manik Surtani  <loonix@users.sourceforge.net>
	[PATCH]
	
	* html/admin/admin_updateDb.php: Fixes SQL issue where undesired directories
	are selected for update (Closes bug #1012797).

2004-11-13  Philip Lowman  <philip@yhbt.com>

	* PQueue/PQueue.pm:
	- Merged in the weighted random algorithm Henrik Wejdmark came up with
	- When instantiating PQueue can now specify government_type &
	random_query_type and appropriate SQL queries are set.
	- government & random_query can be specified to manually override queries
	- timesPlayed is now returned when songs are picked so it can be displayed
	- if an unallowed media type enters the queue it will be removed and an
	error message outputted if it reaches the top of the queue.
	* ices.pm: Modified to use new PQueue.pm options
	* tunezd.pl: Modified to use new PQueue.pm options

2004-11-13  Philip Lowman  <philip@yhbt.com>

	* html/admin/admin_volume.php: Closes patch #1049939 on SourceForge
	although no patch attached.  Fixes issue with last element of
	$_CONF['mixer_devices'] array not showing up on volume page.

2004-09-19  Philip Lowman  <philip@yhbt.com>

	* PQueue/PQueue.pm:
	- Now restricts random song SQL to allowed filetypes
	- Changes allowed_filetypes layout to be a proper hash
	* PQueue/Changes: updated the PQueue changelog
	* tunezd.pl: Uses the new $allowed_filetypes hash
	* ices.pm: Uses the new $allowed_filetypes hash

2004-08-17  Philip Lowman  <philip@yhbt.com>

	* tunezd.pl: More verbose output, eliminated superfluous paramater in
	playsong()

2004-08-17  Philip Lowman  <philip@yhbt.com>

	* ices.pm: Only registers mp3 filetype with PQueue now so OGGs aren't
	returned

2004-08-17  Philip Lowman  <philip@yhbt.com>

	* PQueue/PQueue.pm:
	Added option to module to limit filetypes returned
	Fixed government SQL and random-gen SQL in top() to eliminate the
	possibility of selecting a song which is offline or scheduled for deletion
	Fixed email addresses / URLs

2004-08-17  Philip Lowman  <philip@yhbt.com>

	* PQueue/README: Updated version to 1.21

2004-08-16  Philip Lowman  <philip@yhbt.com>

	* html/admin/admin_updateDb.php: Fixed include as per TbNobody's request
	(assuming it improves Windows compatability)

2004-08-16  CakTux  <caktux@users.sourceforge.net>
	[PATCH]
	
	* html/tunez.inc.php: Added option for setting playInRandomMode en masse
	* html/admin/admin_edit_record.php: HTML to enable above

2004-08-16  Philip Lowman  <philip@yhbt.com>
	
	* config.inc.php: New option for number of seconds to sleep when skipping
	a song
	* html/skip_random.php: Now uses $_CONF['skip_sleepsecs'] for sleep call
	* html/admin/admin_skip.php: Now users $_CONF['skip_sleepsecs'] for sleep
	call

2004-08-16  Henrik Wejdmark  <sepahewe@users.sourceforge.net>
	[PATCH]

	* html/tunez.inc.php: If song length isn't available reload is displayed
	as 1:00 instead of 60

2004-08-16  Philip Lowman  <philip@yhbt.com>

	* html/tunez.inc.php: If nothing is playing timeleft() returns -5000
	which allows header.php to avoid setting a metarefresh
	* html/header.inc.php: If timeleft() returns a value around 0 the metarefresh
	is set to 5 seconds, otherwise it remains the same as before.  If the
	value is way below -5 the metarefresh isn't set at all.

2004-08-16  Henrik Wejdmark  <sepahewe@users.sourceforge.net>
	[PATCH]

	* html/tunez.inc.php: Changed refresh time to 60 seconds for songs without
	known song length.

2004-08-16  Henrik Wejdmark  <sepahewe@users.sourceforge.net>
	[PATCH]

	* html/tunez.inc.php: Fixed absolute image paths

2004-08-16  Philip Lowman  <philip@yhbt.com>

	* Changelog: Moved to ChangeLog.public
	* ChangeLog: Created an actual ChangeLog.


tmixer/src/smixer.c
/*
 * smixer - a simple interface to /dev/mixer
 * Copyright (C) 2000 David Johnson
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 * or view it online at http://www.gnu.org/copyleft/gpl.html
 * 
 */

#define _GNU_SOURCE

#include <stdio.h>
#include <sys/ioctl.h>
#include <sys/soundcard.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <string.h>
#include <unistd.h>
#include <stdlib.h>
#include <getopt.h>

/* defines */
#define SMIXER_VERSION            "v1.0.3"
#define SMIXER_VERSION_DATE       "2003-06-09"

#define SIMXER_MAX_VOL            25700
#define SIMXER_MAX_NAME_LEN       11
#define SMIXER_MIXER_DEV          "/dev/mixer"
#define SMIXER_STR_SIZE           100
#define SMIXER_PLUS_MINUS_AMOUNT  5
#define DELAY                     5000


#define max2(a,b) ( ((a) > (b)) ? (a) : (b) )
#define min2(a,b) ( ((a) < (b)) ? (a) : (b) )
#define max3(a,b,c) ( max2((a),max2((b),(c))) )
#define min3(a,b,c) ( min2((a),min2((b),(c))) )
#define between(min,a,max) ( max2(min2((a),(max)),(min)) )



/* globals */
const char *smixer_dev_names_const[] = SOUND_DEVICE_LABELS;
char smixer_dev_names[SOUND_MIXER_NRDEVICES][SIMXER_MAX_NAME_LEN+1];

/* structs */
typedef struct
{
  int devmask;
  int recsrc;
  int recmask;
  int stereo;
  int vols[SOUND_MIXER_NRDEVICES];
} smixer_data_t;

/* functions */
static void smixer_remove_spaces_and_caps(void);
static void smixer_read_mixer(int do_all);
static void simxer_write_mixer(char *filename, int do_all);
static int smixer_get_data(int fd, smixer_data_t *data);
static void smixer_print_data(const smixer_data_t *data, int do_all);
static char* smixer_nocaps (char *text);
static void smixer_print_help(void);

/* main */

int main(int argc, char *argv[])
{
  int do_p = 0;
  int do_s = 0;
  int do_a = 0;
  int ok = 1;
  char *filename = NULL;
  int retval;
  const struct option longopts[] = 
    {
      { "print",  0, NULL, 'p' },
      { "all",    0, NULL, 'a' },
      { "set",    2, NULL, 's' },
      { 0 }
    };
  
  while ( 1 )
    {
      retval = getopt_long(argc, argv, "pas::", longopts, NULL);
      
      if (retval < 0)
        {
          break;
        }
      
      switch (retval)
        {
        case 'p':
          do_p = 1;
          break;
          
        case 'a':
          do_a = 1;
          break;
          
        case 's':
          do_s = 1;
          filename = optarg;
          break;
          
        default:
          ok = 0;
          break;
        }
    }

  if (optind < argc)
    {
      ok = 0;
    }
  
  smixer_remove_spaces_and_caps();
  
  if ( ok && do_p && !do_s )
    {
      smixer_read_mixer(do_a);
    }
  else if ( ok && do_s && !do_p )
    {
      simxer_write_mixer(filename, do_a);
    }
  else
    {
      fprintf(stderr,
              "smixer " SMIXER_VERSION " (" SMIXER_VERSION_DATE ") Copyright (C) 2000-2003 David Johnson\n"
              "  http://centerclick.org/programs/smixer/\n"
              "  smixer@centerclick.org\n"
              "usage: %s [-a] -p           (print settings)\n"
              "usage: %s [-a] --set [file] (set settings from file or stdin)\n",
              argv[0],argv[0]);
      return (1);
    }
  
  return (0);
}


static void smixer_read_mixer(int do_all)
{
  int mixer_fd;
  smixer_data_t data;
  
  if ( ( mixer_fd = open(SMIXER_MIXER_DEV,O_RDWR) ) < 0 )
    {
      perror("open " SMIXER_MIXER_DEV);
      goto error_out;
    }
  
  if ( smixer_get_data(mixer_fd,&data) < 0 )
    {
      goto error_out;
    }
  
  smixer_print_data(&data, do_all);
  
 error_out:
  
  if (mixer_fd > 2)
    {
      close(mixer_fd);
    }
  
  return;
}

static void simxer_write_mixer(char *filename, int do_all)
{
  int step;
  int mixer_fd=0, i, num_args, found;
  FILE *config_file=NULL;
  smixer_data_t data;
  char templine[SMIXER_STR_SIZE];
  char args[3][SMIXER_STR_SIZE];
  int vol1, vol2;
  
  if ( !filename || !strcmp(filename,"-") )
    {
      config_file = stdin;
    }
  else
    {
      if ( ( config_file = fopen(filename,"r") ) == NULL )
        {
          perror("open config");
          goto error_out;
        }
    }
  
  fflush(stdout);
  
  if ( ( mixer_fd = open(SMIXER_MIXER_DEV,O_RDWR) ) < 0 )
    {
      perror("open " SMIXER_MIXER_DEV);
      goto error_out;
    }
  
  while ( fgets(templine,SMIXER_STR_SIZE,config_file) )
    {
      
      memset(args[0],0,SMIXER_STR_SIZE);
      memset(args[1],0,SMIXER_STR_SIZE);
      vol1 = -1;
      vol2 = -1;
      
      num_args = sscanf(templine,"%s %s %s\n",args[0],args[1],args[2]);
      
      if ( num_args == 0 || args[0][0] == '\0' || args[0][0] == '#' )
        {
          continue;
        }
      
      if ( args[0][0] == 'v' )
        {
          found=0;
          smixer_nocaps(args[1]);
          
          if (num_args < 3)
            {
              fprintf(stderr,"vol: missing argument\n");
              continue;
            }
          
          if ( smixer_get_data(mixer_fd,&data) < 0 )
            {
              goto error_out;
            }
          
          for (i=0; i<SOUND_MIXER_NRDEVICES; i++)
            {
              if ( data.devmask & (1 << i) && !strcmp(args[1],smixer_dev_names[i]) )
                {
                  found++;
                  
                  if ( !strcmp(args[2],"+") ) /* + 5 % */
                    {
                      vol1 = between(0, data.vols[i] + (SMIXER_PLUS_MINUS_AMOUNT*SIMXER_MAX_VOL/100), SIMXER_MAX_VOL);
                    }
                  else if ( !strcmp(args[2],"-") ) /* - 5 % */
                    {
                      vol1 = between(0, data.vols[i] - (SMIXER_PLUS_MINUS_AMOUNT*SIMXER_MAX_VOL/100), SIMXER_MAX_VOL);
                    }
                  else
                    {
                      vol1 = (int)((float)SIMXER_MAX_VOL * (float)atoi(args[2]) / 100.0);
                    }
                  
                  if (vol1 < 0 || vol1 > SIMXER_MAX_VOL)
                    {
                      fprintf(stderr,"vol: value out of range\n");
                      continue;
                    }
                  vol2 = (data.vols[i]&0xff) + ((data.vols[i]&0xff)<<8);
                
                  if (vol1 < vol2) {
                    step = -0x101;
                  }
                  else { //if (vol2 < vol1) {
                    step = 0x101;
                  }
                
                  while (vol1 != vol2) {
                      vol2 += step;
                      if (ioctl(mixer_fd,MIXER_WRITE(i),&vol2) < 0) {
                          perror("write vol");
                          goto error_out;
                      }
                      usleep(DELAY);
                  }
                  data.vols[i] = vol2;

                }
            }
          
          if (!found)
            {
              fprintf(stderr,"vol: name not found: %s\n",args[1]);
            }
        }
      
      else if ( args[0][0] == 'r' )
        {
          found=0;
          smixer_nocaps(args[1]);
          
          if ( smixer_get_data(mixer_fd,&data) < 0 )
            {
              goto error_out;
            }
          
          for (i=0; i<SOUND_MIXER_NRDEVICES; i++)
            {
              if ( data.recmask & (1 << i) && !strcmp(args[1],smixer_dev_names[i]) )
                {
                  found++;
                  vol1 = (1 << i);
                  
                  if ( ioctl(mixer_fd,SOUND_MIXER_WRITE_RECSRC,&vol1) < 0 )
                    {
                      perror("write vol");
                      goto error_out;
                    }
                  
                  if ( ioctl(mixer_fd,SOUND_MIXER_READ_RECSRC,&vol2) < 0 )
                    {
                      perror("comfirm vol");
                      goto error_out;
                    }
                  
                  data.recsrc = vol2;
                  
                  if ( vol1 != vol2 )
                    {
                      fprintf(stderr,
                              "comparision failed: tried 0x%.8X got 0x%.8X\n",
                              vol1,vol2);
                    }
                }
            }
          
          if (!found)
            { 
              fprintf(stderr,"recsrc: name not found: %s\n",args[1]);
            }
  
        }

      else if ( args[0][0] == 's' )
        {
          if ( smixer_get_data(mixer_fd,&data) < 0 )
            {
              goto error_out;
            }
          
          smixer_print_data(&data, do_all);
        }
     
      else if ( (args[0][0] == 'h') || (args[0][0] == '?') )
        {
          smixer_print_help();
        }
     
      else if ( (args[0][0] == 'q') || (args[0][0] == 'e') )
        {
          goto error_out;
        }
     
      else
        {
          fprintf(stderr,"unknown command try \"help\": %s\n",args[0]);
        }
     
    }
 
 error_out:
 
  if (mixer_fd > 2) { close(mixer_fd); }
  if (config_file && (config_file != stdin)) { fclose(config_file); }
  
  return;
}

static int smixer_get_data (int fd, smixer_data_t *data)
{
  int i;
  
  memset(data,0,sizeof(smixer_data_t));
  
  if ( ioctl(fd,SOUND_MIXER_READ_DEVMASK,&data->devmask) < 0 )
    {
      perror("read SOUND_MIXER_DEVMASK");
      return -1;
    }
  
  if ( ioctl(fd,SOUND_MIXER_READ_RECSRC,&data->recsrc) < 0 )
    {
      perror("read SOUND_MIXER_RECSRC");
      return -1;
    }
  
  if ( ioctl(fd,SOUND_MIXER_READ_RECMASK,&data->recmask) < 0 )
    {
      perror("read SOUND_MIXER_RECMASK");
      return -1;
    }
  
  if ( ioctl(fd,SOUND_MIXER_READ_STEREODEVS,&data->stereo) < 0 )
    {
      perror("read SOUND_MIXER_STEREODEVS");
      return -1;
    }
  
  for (i=0; i<SOUND_MIXER_NRDEVICES; i++)
    {
      if ( data->devmask & ( 1 << i) )
        {
          if ( ioctl(fd,MIXER_READ(i),&data->vols[i]) < 0 )
            {
              perror("read SOUND_MIXER_STEREODEVS");
              return -1;
            }
        }
    }
  
  return 0;
}

static void smixer_print_data (const smixer_data_t *data, int do_all)
{
  int i;
  
  printf("ID: Name      Play  Rec   Channels  Volume\n"
         "--  --------  ----  ----  --------  ------\n");
  
  for (i=0; i<SOUND_MIXER_NRDEVICES; i++)
    {
      if ( (data->devmask | data->recmask) & (1 << i) )
        {
          printf("%2i: %-8s  %-4s  %-4s  %-8s  %5.0f%%\n",
                 i,
                 smixer_dev_names_const[i],
                 (data->devmask & (1 << i)) ? (data->vols[i] ? "on" : "off" ) : "-",
                 (data->recmask & (1 << i)) ? ((data->recsrc & (1 << i)) ? "on" : "off" ) : "-",
                 (data->stereo  & (1 << i)) ? "stereo" : "mono",
                 100.0*((float)data->vols[i])/((float)SIMXER_MAX_VOL));
        }
      else if (do_all)
        {
          printf("%2i: %-8s  -     -     -              -\n",i,smixer_dev_names_const[i]);
        }
    }
  return;
}


static void smixer_print_help(void)
{
  
  printf("\nsmixer commands:\n"

         "\nvol [name] [value|-|+]\n"
         "  'vol' sets the volume for a specific input or output device\n"
         "  'name' is the name of the device, do smixer -p to get a list\n"
         "  'value' is the percentage volume (no %%)\n"
         "    or \"-\" to decrease 5%% or \"+\" to increase 5%%\n"
 
         "\nrecsrc [name]\n"
         "  sets the recording source to a specific input device\n"
         "  'name' is the name of the device, do smixer -p to get a list\n"
 
         "\nshow\n"
         "  prints the same list as 'smixer -p' does\n"
         "\nhelp|?\n"
         "  gets help\n"

         "\nend|exit|quit|q\n"
         "  exits\n"

         );
  return;
}


static void smixer_remove_spaces_and_caps(void)
{
  int i,j;
  
  for (i=0; i<SOUND_MIXER_NRDEVICES; i++)
    {
      strncpy(smixer_dev_names[i],smixer_dev_names_const[i],SIMXER_MAX_NAME_LEN);
      smixer_nocaps(smixer_dev_names[i]);
      for (j=0; smixer_dev_names[i][j]; j++)
        {
          if ( smixer_dev_names[i][j] == ' ')
            {
              smixer_dev_names[i][j] = '\0';
            }
        }
    }
  return;
}

static char* smixer_nocaps(char *text)
{
  int i;
  
  if (text)
    {
      for (i=0; text[i]; i++)
        {
          if ( text[i] >= 'A' && text[i] <= 'Z' )
            {
              text[i] += 'a' - 'A';
            }
        }
    }
  return text;
}
